interceptor/zipkin/trace_interceptor.go

```diff
@@ -0,0 +1,399 @@
+	return nil
+}
+
+func (zi *ZipkinInterceptor) parseAndConvertToTraceSpans(jsonBlob []byte) (reqs []*agenttracepb.ExportTraceServiceRequest, err error) {
+	var zipkinSpans []*zipkinmodel.SpanModel
+	if err = json.Unmarshal(jsonBlob, &zipkinSpans); err != nil {
+		return nil, err
+	}
+
+	// *commonpb.Node instances have unique addresses hence
+	// for grouping within a map, we'll use the .String() value
+	byNodeGrouping := make(map[string][]*tracepb.Span)
+	nodes := make([]*commonpb.Node, 0, len(zipkinSpans))
+	// Now translate them into tracepb.Span
+	for _, zspan := range zipkinSpans {
+		span, node, err := zipkinSpanToTraceSpan(zspan)
+		// TODO:(@odeke-em) record errors
+		if err == nil && span != nil {
+			key := node.String()
+			if _, firstTime := byNodeGrouping[key]; firstTime {
+				nodes = append(nodes, node)
+			}
+			byNodeGrouping[key] = append(byNodeGrouping[key], span)
+		}
+	}
+
+	// TODO: @odeke-em perhaps maintain the relative order that the spans were received in
+	for _, node := range nodes {
+		key := node.String()
+		spans := byNodeGrouping[key]
+		if len(spans) == 0 {
+			// Should never happen but nonetheless be cautious
+			// not to send blank spans.
+			continue
+		}
+		reqs = append(reqs, &agenttracepb.ExportTraceServiceRequest{
+			Node:  node,
+			Spans: spans,
+		})
+		delete(byNodeGrouping, key)
+	}
+
+	return reqs, nil
+}
+
+func (zi *ZipkinInterceptor) StopTraceInterception(ctx context.Context) error {
+	return nil
+}
+
+// processBodyIfNecessary checks the "Content-Encoding" HTTP header and if
+// a compression such as "gzip", "deflate", "zlib", is found, the body will
+// be uncompressed accordingly or return the body untouched if otherwise.
+// Clients such as Zipkin-Java do this behavior e.g.
+//    send "Content-Encoding":"gzip" of the JSON content.
+func processBodyIfNecessary(req *http.Request) io.Reader {
+	switch req.Header.Get("Content-Encoding") {
+	default:
+		return req.Body
+
+	case "gzip":
+		return gunzippedBodyIfPossible(req.Body)
+
+	case "deflate", "zlib":
+		return zlibUncompressedbody(req.Body)
+	}
+}
+
+func gunzippedBodyIfPossible(r io.Reader) io.Reader {
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		// Just return the old body as was
+		return r
+	}
+	return gzr
+}
+
+func zlibUncompressedbody(r io.Reader) io.Reader {
+	zr, err := zlib.NewReader(r)
+	if err != nil {
+		// Just return the old body as was
+		return r
+	}
+	return zr
+}
+
+// The ZipkinInterceptor receives spans from endpoint /api/v2 as JSON,
+// unmarshals them and sends them along to the spanreceiver.
+func (zi *ZipkinInterceptor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Trace this method
+	ctx, span := trace.StartSpan(context.Background(), "ZipkinInterceptor.Export")
+	defer span.End()
+
+	// If the starting RPC has a parent span, then add it as a parent link.
+	parentCtx := r.Context()
+	internal.SetParentLink(parentCtx, span)
+
+	pr := processBodyIfNecessary(r)
+	slurp, err := ioutil.ReadAll(pr)
```

Comment:
It seems advisable to defend against bad data here.

Reply:
What do you mean? We read all the data into memory and then JSON unmarshal it. Are you suggesting say a max limit? If so, net/http already handles that and has tests about content-length mismatches.

Reply:
My assumption about the "decompressing" readers gave the impression that decompression could fail here.